test/test_genome.py

```diff
@@ -346,89 +347,118 @@ def test_is_gene_in_output_region(rng_seed):
+    n_column = 4
+    n_row = 3
+    genome = cgp.Genome(n_input, n_output, n_column, n_row, None, (cgp.Add, cgp.Sub))
     rng = np.random.RandomState(rng_seed)
-    genome = cgp.Genome(1, 1, 3, 1, None, (cgp.Add, cgp.ConstantFloat))
-    dna = [
-        ID_INPUT_NODE,
-        ID_NON_CODING_GENE,
-        ID_NON_CODING_GENE,
-        1,
-        0,
-        0,
-        1,
-        0,
-        0,
-        0,
-        0,
-        2,
-        ID_OUTPUT_NODE,
-        3,
-        ID_NON_CODING_GENE,
-    ]
-    genome.dna = list(dna)
-    active_regions = cgp.CartesianGraph(genome).determine_active_regions()
+    genome.randomize(rng)
 
-    # mutating any gene in inactive region returns True
-    assert genome._mutate_hidden_region(list(dna), 3, active_regions, rng) is True
-    assert genome._mutate_hidden_region(list(dna), 4, active_regions, rng) is True
-    assert genome._mutate_hidden_region(list(dna), 5, active_regions, rng) is True
+    def count_n_immutable_genes(n_input, n_output, n_row):
+        counter = n_input * 3  # 3 is the arity
+        counter += n_output * 2  # 2 of 3 output genes can not be mutated
+        if n_input == 1:
+            counter += (
+                n_row * 2
+            )  # input gene addresses in the first hidden layer can't be mutated in that case
+        return counter
+
+    def count_mutations(prev_dna, new_dna):
+        counter = 0
+        for idx in range(len(prev_dna)):
+            if prev_dna[idx] != new_dna[idx]:
+                counter += 1
+        return counter
+
+    acceptable_error_interval = 0.05  # Todo: get comment for this value from Jakob and Max
+    n = 1000
+    n_immutable_genes = count_n_immutable_genes(n_input, n_output, n_row)
+    n_mutations_expected = n * mutation_rate * (len(genome.dna) - n_immutable_genes)
+    print(len(genome.dna), n_immutable_genes)
+    n_mutations = 0
+    for _ in range(n):
+        temp_dna = genome.dna
+        genome.mutate(mutation_rate, rng)
+        current_n_mutations = count_mutations(temp_dna, genome.dna)
+        n_mutations += current_n_mutations
+
+    assert abs(n_mutations_expected - n_mutations) < (
+        n_mutations_expected * acceptable_error_interval
+    )
 
-    # mutating function gene in active region returns False
-    assert genome._mutate_hidden_region(list(dna), 6, active_regions, rng) is False
-    # mutating inactive genes in active region returns True
-    assert genome._mutate_hidden_region(list(dna), 7, active_regions, rng) is True
-    assert genome._mutate_hidden_region(list(dna), 8, active_regions, rng) is True
 
-    # mutating any gene in active region without silent genes returns False
-    assert genome._mutate_hidden_region(list(dna), 9, active_regions, rng) is False
-    assert genome._mutate_hidden_region(list(dna), 10, active_regions, rng) is False
-    assert genome._mutate_hidden_region(list(dna), 11, active_regions, rng) is False
+def test_only_silent_mutations(genome_params, rng_seed):
```

Comment:
can we somehow test that mutating specific genes returns the correct value for `only_silent_mutations` in addition to this brute force approach?

Reply:
It is tricky, since the indices selected for mutation `selected_gene_indices` are calculated locally within `genome.mutate`. Is there maybe a way to overwrite a local variable with a global one in pytest? (couldn't find anything with a quick search in the docs...)

Reply:
i see, that is a bit tricky. two options come to mind: i) adding a function to `Genome` that returns the `selected_gene_indices` and monkey patching it in the test to return the gene indices that you'd like to use or ii) split `mutate` at the line where `selected_gene_indices` is set such that `Genome` has a mutate function that operates on a list of gene indices. i think i would go for i) for now and after this PR has been merged think about how we can refactor `mutate` to make it simpler to understand and test